{"cells":[{"cell_type":"markdown","metadata":{"id":"PQjn3YPmxQDj"},"source":["# Model\n","\n","In this script we are trying to build the model to validate our idea. \n","We will take the data from the bulk analysis as training data, and generate a matrix with fake proportion of cells just to train the model. \n","For the final script model the bulk data can then be easily replaced with single-cell data, and at the bottom of the script  we just need to add the bulk data to predict the proportion of cells.\n","\n","In Google Colab set the runtine as: ==R with TPU==; running it without TPU resulted in errors (possibly solved with \"reticulate\" package, code commented).\n","\n","File(s) needed: \n","    - 'bulk_tissue_data.csv' from notebook `01scripts\\01_compBiology_bulk.ipynb`\n","\n","## load libraries and functions"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":121962,"status":"ok","timestamp":1694683996838,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"EIUGDGU9xK8x","outputId":"a42408e5-87d7-4d61-dc3b-a5011df3d81f","vscode":{"languageId":"r"}},"outputs":[],"source":["# libraries ####\n","#devtools::install_github(\"rstudio/keras\")\n","#install.packages(\"tidyverse\")\n","# install.packages(\"gradDescent\")\n","library(keras)\n","library(tidyverse)\n","#library(gradDescent)\n","\n","## custom functions ####\n","# create a function to generate mock data about the proportion of each cell type. For this the sum of the is 1\n","# with this function with can specify the mock number of cell type by the `n_rows` and the number of samples as `n_cols`\n","\n","create_matrix <- function(n_rows, n_cols) {\n","  # Generate a vector of random numbers between 0 and 1\n","  data <- runif(n_rows * n_cols)\n","\n","  # Create a matrix from the vector\n","  matrix <- matrix(data, nrow = n_rows, ncol = n_cols)\n","\n","# Normalize the matrix such that the sum of each column is 1\n","  for (i in 1:n_cols) {\n","    matrix[, i] <- matrix[, i] / sum(matrix[, i])\n","  }\n","\n","  return(matrix)\n","}\n","\n","# as alternative scaling function --- NOT USED\n","min_max_scaling <- function(x) {\n","  return ((x - min(x)) / (max(x) - min(x)))\n","}"]},{"cell_type":"markdown","metadata":{},"source":["## load data\n","\n","In the `csv` data the first column (features) has the genes and other features such as `age` and `condition`. However, we did not changed the column name `ensembl_gene_id`.\n","To scale the data (standardization)  so that we can use data from bulk and single-cell RNA-seq, we need to split this data between what are genes and what is not, because we just want to scale the genes. The other variables are already numeric or need to be converted to  binary."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":592},"executionInfo":{"elapsed":11252,"status":"ok","timestamp":1694684314982,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"eStcg3b2xinK","outputId":"d8190626-502a-41dd-af98-0a7028eaeddb","vscode":{"languageId":"r"}},"outputs":[],"source":["# load data ####\n","tissue_bulk_data <- read.csv(\"bulk_tissue_data.csv\")\n","data <- tissue_bulk_data #[c(1:5,c(17276:17279)),1:5]\n","tail(data)\n","## preprocess data  ####\n","split_gene_features <- data[c(1:17275),-1]\n","tail(split_gene_features)\n"]},{"cell_type":"markdown","metadata":{},"source":["From this table (above) we can see that we have exported more features that what we need to be able to use the two data sets. For example,\"condition_specified\" is not present in single-cell-RNAseq data. Also, the sex M/F, so we need to convert to 0/1. Therefore, this is information that we can ignore in this script."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":118},"executionInfo":{"elapsed":250,"status":"ok","timestamp":1694684319369,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"yNxv57_95g50","outputId":"1aae25ed-d69d-4837-e4a0-b8f538f72052","vscode":{"languageId":"r"}},"outputs":[],"source":["### convert to binary\n","binary_sex <- ifelse(data[17279,-1 ] == \"F\", 1, 0)\n","binary_condition <- ifelse(data[17276,-1 ] == \"control\", 0, 1)\n","binary_sex"]},{"cell_type":"markdown","metadata":{},"source":["Now with the split feature genes from other phenoData we can scale the genes using the function `scale()`. We decided to use this standard function, but it is something that we can research and possibbly find better option for this  kind of genomic data. "]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":274},"executionInfo":{"elapsed":16842,"status":"ok","timestamp":1694684339559,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"9O7fKoA96dfF","outputId":"15b2a2c5-2a8f-4837-b489-99c2ea7e1cb3","vscode":{"languageId":"r"}},"outputs":[],"source":["scale_genes <- t(scale(t(data.matrix(split_gene_features))))\n","head(scale_genes)\n","# merge scale genes with other feature (sex and condition)\n","x_original <- rbind(scale_genes,binary_sex,binary_condition)"]},{"cell_type":"markdown","metadata":{},"source":["Now let's just print the original data to check that the `scale()` worked."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":274},"executionInfo":{"elapsed":302,"status":"ok","timestamp":1694684346956,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"qGkrJPvl60xv","outputId":"9fdb08e9-1c09-42e3-9e21-04491941712c","vscode":{"languageId":"r"}},"outputs":[],"source":["tail(x_original)"]},{"cell_type":"markdown","metadata":{},"source":["It worked. Now we have our `x` data ready to build our model, but since we don't have the proportion of the cell type for the bulk data we will generate mock that with the function define above. \n","For sake of simplicity we will only consider 4 cell types, and, naturally, the number of samples must the same (aka the number of patient samples)."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":232,"status":"ok","timestamp":1694684349590,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"njDgtdTnxp2G","vscode":{"languageId":"r"}},"outputs":[],"source":["# Create a mock matrix for cell type porportion with random numbers\n","number_of_cell_types = 4\n","number_of_samples = dim(x_original)[2] # number of patients\n","y_original <- create_matrix(number_of_cell_types,number_of_samples)\n","\n","# Normalize the columns to sum to 1\n","y_original <- y_original / colSums(y_original)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":180},"executionInfo":{"elapsed":274,"status":"ok","timestamp":1694684353970,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"WLriISMb4IPP","outputId":"b11d4925-5ea2-48e6-887c-7b730104e3ea","vscode":{"languageId":"r"}},"outputs":[],"source":["y_original"]},{"cell_type":"markdown","metadata":{},"source":["# Build model\n","\n","Ok, now we have our data ready to build the model. For that we will used keras package, and as starting point we will consider that the number of features to be equal to number of genes; this might need to be optimized. \n","But, before we proceed, let's just check that the matrices size are correct."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":69},"executionInfo":{"elapsed":233,"status":"ok","timestamp":1694684356631,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"ggAxFM6gxs4J","outputId":"5f4011eb-570e-48c1-8db6-444b8cfc4975","vscode":{"languageId":"r"}},"outputs":[],"source":["# build the model ####\n","## define feature and targets \n","x <- x_original\n","y <- y_original\n","# check matrix dims\n","tissue_composition <- dim(y)[1]\n","number_of_features <- dim(x)[1]\n","dim(x)\n","dim(y)\n","tissue_composition"]},{"cell_type":"markdown","metadata":{},"source":["We can now see the the matrix is has the samples in rows, so we need to transpose the matrix to build the model; features should be the columns and rows the observation (patient samles).\n","At this stage we also need to split that data into \"training\" and \"testing\"; we will consider arbitrary proportion of  $0.8$ for training and $0.2$ for testing."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":630,"status":"ok","timestamp":1694684364380,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"ylX71AhTBHAk","vscode":{"languageId":"r"}},"outputs":[],"source":["# Assuming you have a dataset 'data' with features 'x' and labels 'y'\n","x <- t(x)\n","y <- t(y)\n","# Set a random seed for reproducibility\n","set.seed(123)\n","\n","# Define the proportion of data to use for testing (e.g., 20%)\n","test_split_ratio <- 0.2\n","\n","# Generate random indices for splitting the data\n","num_samples <- nrow(x)\n","num_test_samples <- round(num_samples * test_split_ratio)\n","test_indices <- sample(1:num_samples, num_test_samples)\n","\n","# Split the data into training and test sets\n","x_train <- x[-test_indices, ]  # Training features\n","y_train <- y[-test_indices, ]  # Training labels\n","x_test <- x[test_indices, ]    # Test features\n","y_test <- y[test_indices, ]    # Test labels"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":7340,"status":"ok","timestamp":1694684375500,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"cj3F8q4n7W3J","vscode":{"languageId":"r"}},"outputs":[],"source":["# Initialize model\n","# reticulate::use_condaenv(\"base\", conda = \"auto\") # run on cpu\n","model <- keras_model_sequential()\n"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":427,"status":"ok","timestamp":1694684383214,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"oTOUKXwi8TFD","vscode":{"languageId":"r"}},"outputs":[],"source":["# Add layers\n","feature_factor = 1\n","model %>%\n","  layer_dense(units = feature_factor*number_of_features, activation = 'sigmoid', input_shape = dim(x)[2]) %>% # units=265 relu\n","  layer_dropout(rate = 0.4) %>%\n","  layer_dense(units = 128, activation = 'sigmoid') %>% #units = 128 relu\n","  layer_dropout(rate = 0.3) %>%\n","  layer_dense(units = tissue_composition, activation = 'softmax')\n"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":263,"status":"ok","timestamp":1694684399833,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"-CM9DCbs8Pxv","vscode":{"languageId":"r"}},"outputs":[],"source":["# Compile model\n","model %>% compile(\n","  loss = 'categorical_crossentropy', # binary_crossentropy #\n","  optimizer = optimizer_rmsprop(),#optimizer_rmsprop(),\n","  metrics = c('accuracy')\n",")"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":31472,"status":"ok","timestamp":1694684489476,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"TeLIs5BQ8j9k","vscode":{"languageId":"r"}},"outputs":[],"source":["# Train model\n","history <- model %>% fit(\n","  x_train,\n","  y_train,\n","  epochs = 100, #30\n","  batch_size = 32, # 32\n","  #validation_split = 0.2\n",")"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":437},"executionInfo":{"elapsed":884,"status":"ok","timestamp":1694684493588,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"ZAZAKqOru29s","outputId":"9cdec969-ce16-4d7e-b6e1-e68dcc112f7b","vscode":{"languageId":"r"}},"outputs":[],"source":["plot(history)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"elapsed":600,"status":"ok","timestamp":1694684654190,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"1syf-Zwn8hhv","outputId":"6a6f3b55-740d-4dd5-8e90-14dd3d48ecd9","vscode":{"languageId":"r"}},"outputs":[],"source":["# Evaluate model\n","model %>% evaluate(x_test, y_test)\n","\n","# Make predictions\n","predictions <- model %>% predict(x_test) %>% k_argmax()"]},{"cell_type":"markdown","metadata":{},"source":["The code seems to be running without errors. We can see that the accuracy is not great, but this is expected since we used random data to train the model.  \n","Now we just want to check that the results for the predictions have a sum closer to 1. If the sum is not 1 it might indicate that the last function is not properly applied. "]},{"cell_type":"code","execution_count":null,"metadata":{"vscode":{"languageId":"r"}},"outputs":[],"source":["predictions = model %>% predict(x_train)\n","head(predictions)\n","head(model %>% predict(x_train)) %>% apply(1, sum)"]},{"cell_type":"markdown","metadata":{},"source":["Good, the sum is closer to one. Now let's just check how different are the values predict form the training values (yet knowing that the accuracy is not great)."]},{"cell_type":"code","execution_count":null,"metadata":{"vscode":{"languageId":"r"}},"outputs":[],"source":["head(y_train)\n","head(y_train %>% apply(1, sum))"]},{"cell_type":"markdown","metadata":{},"source":["Ok, great that is it! The code for the model is running."]},{"cell_type":"markdown","metadata":{},"source":["# Prepare bulk data to be used by model trained with single-cell-RNAseq data"]},{"cell_type":"markdown","metadata":{},"source":["The steps here are identical to what was performed before, but have some adjustments like cheking that the genes between datasets are the same and in the same order in the table. \n","We decided to do this step because in the first layer the number of features is the same as the number of feature, therefore we thought that their order could have implications. Thus, to avoid adding unnecessary variables we made sure that the features are equally ordered (aka have the same index)."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":484},"executionInfo":{"elapsed":3787,"status":"ok","timestamp":1694630761812,"user":{"displayName":"Leonardo Mendes-Silva","userId":"04125321159063099921"},"user_tz":-60},"id":"NTPD4SN1xGHf","outputId":"740e42ab-7fb9-47d9-f926-b66bb1f6f877","vscode":{"languageId":"r"}},"outputs":[],"source":["# process and prepare bulk for pipeline\n","data_bulk <- read.csv(\"bulk_tissue_data.csv\")\n","genes_in_singleCell <- tissue_bulk_data$ensembl_gene_id[c(11,5,9,8)] #random rows, replace with genes from csv from single-cell\n","genes_filter_bulk <- data_bulk %>% filter(ensembl_gene_id %in% genes_in_singleCell)\n","matched_gene_order_bulk <- genes_filter_bulk %>% arrange(match(ensembl_gene_id,genes_in_singleCell))\n","# remove ensembl\n","split_gene_features <- matched_gene_order_bulk[,-1]\n","### convert to binary\n","binary_sex <- ifelse(data_bulk[17276,] == \"F\", 1, 0)[,-1]\n","binary_condition <- ifelse( data_bulk[17279,] == \"control\", 0, 1)[,-1]\n","# scale\n","scale_genes <- t(scale(t(data.matrix(split_gene_features))))\n","head(scale_genes)\n","head(binary_sex)\n","head(binary_condition)\n","bulk_genes_filtered <- rbind(scale_genes,binary_sex,binary_condition)\n","bulk_genes_filtered"]}],"metadata":{"accelerator":"GPU","colab":{"gpuType":"T4","provenance":[]},"kernelspec":{"display_name":"R","language":"R","name":"ir"},"language_info":{"codemirror_mode":"r","file_extension":".r","mimetype":"text/x-r-source","name":"R","pygments_lexer":"r","version":"4.0.5"}},"nbformat":4,"nbformat_minor":0}
